<!DOCTYPE html>
<html>
<head>
  <meta charset='utf-8'>

  <title>fls-bioinformatics-core/RnaChipIntegrator @ GitHub</title>

  <style type="text/css">
    body {
      margin-top: 1.0em;
      background-color: #ffffff;
      font-family: Helvetica, Arial, FreeSans, san-serif;
      color: #000000;
    }
    #container {
      margin: 0 auto;
      padding: 20px;
      width: 800px;
      background: #ffffff;
    }
    h1 { font-size: 3.0em; color: #000000; margin-bottom: 3px; }
    h1 .small { font-size: 0.4em; }
    h1 a { text-decoration: none }
    h2 { font-size: 1.2em; color: #000000; }
    h3 { text-align: center; color: #000000; }
    a { color: #000000; }
    .description { font-size: 1.1em; margin-bottom: 20px; margin-top: 10px; font-style: italic;}
    .download { float: right; }
    pre { background: #000; color: #fff; padding: 15px;}
    hr { border: 0; width: 80%; border-bottom: 1px solid #aaa}
    .footer { text-align:center; padding-top:30px; font-style: italic; }
    .box { padding:10px; background: #ddd; }
  </style>
</head>

<body>

  <div id="container">

    <h1><a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator">RnaChipIntegrator</a></h1>

    <div class="description">
      Integrated analyses of RNA-Seq and other genomic feature data with ChIP-Seq data
    </div>

    <div class="box">
    <h2>About</h2>
      <p>RnaChipIntegrator is a utility that performs integrated analyses of RNA-Seq and ChIP-Seq
	data, identifying the nearest ChIP peaks to each transcript, and vice versa. The individual
	analyses differ from each other according to the criteria that are used to calculate the
	distances between peaks and transcripts.</p>

      <p>The program was originally written specifically for ChIP-Seq and RNA-Seq data but can be
	used with any set of genomic features e.g. canonical genes, CpG islands etc, or expression
	data e.g. microarrays.</p>
    </div>
    
    <h2>Download</h2>

      <p>Download the latest version:</p>

      <ul>
	<li><a href="https://github.com/downloads/fls-bioinformatics-core/RnaChipIntegrator/RnaChipIntegrator-0.3.2.tar.gz">RnaChipIntegrator-0.3.2.tar.gz</a></li>
	<li><a href="https://github.com/downloads/fls-bioinformatics-core/RnaChipIntegrator/RnaChipIntegrator-0.3.2.zip">RnaChipIntegrator-0.3.2.zip</a></li>
      </ul>

       See the <a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator/downloads">downloads page</a> for all available versions.</p>
    
    <h2>Installation</h2>

      <p>You can use Python's setuptools to install the RnaChipIntegrator code: download the
	archive (either .tar.gz or .zip), unpack it and then do:</p>

      <pre>python setup.py install</pre>

      <p>in the unpacked directory.</p>

      <p>Alternatively if you're using pip to manage your Python installations then you can do</p>

      <pre>pip install RnaChipIntegrator.tar.gz</pre>

      <p>If you don't want to do a system install then you can just unpack the archive and run
	the program directory using:</p>

      <pre>/path/to/RnaChipIntegrator.py [options] ...</pre>

    <h2>Usage</h2>

      <p>The most basic form of usage is:</p>

      <pre>RnaChipIntegrator &lt;rna-data-file&gt; &lt;chip-data-file&gt;</pre>

      <p>The RNA-seq data file must be a tab-delimited file with 5 columns of data for each genomic
	feature (one per line):</p>

      <pre># ID  chr  start  end  strand</pre>

      <p>chr is the chromosome name, start and end define the limits of the feature, and strand
	must be either + or -. ID is a name which is used to identify the genomic feature in the
	output.</p>

      <p>Optionally there can be a 6th column, indicating whether the gene was differentially
	expressed (= 1) or not (= 0).</p>

      <p>The ChIP-seq data file must be a tab-delimited file with 3 columns of data for each
	ChIP peak (one per line):</p>

      <pre># chr  start  end</pre>

      <p>chr is the chromosome name (must match those in the RNA-seq file), and start and end
	define the ChIP peaks - these can either be summits (in which case 'end' - 'start' = 1),
	or regions (with 'start' and 'end' indicating the extent).</p>

      <p>Note that different analyses will be selected depending on whether the ChIP peaks are
	defined as summits or regions.</p>

      <p>Also running</p>

      <pre>RnaChipIntegrator.py -h</pre>

      <p>will produce a list of options and descriptions of their functions.</p>
    
      <h2>Dependencies</h2>
      <p>RnaChipIntegrator was written for Python 2.7, but has also been used with Python
	2.4. It uses the xlwt, xlrd and xlutils Python libraries to write its XLS output files.
	If these are not available then the program will still run but won't produce an XLS
	file.</p>
    
      <h2>License</h2>
      <p>Artistic Licence 2.0</p>
    
      <h2>Authors</h2>
      <p>Ian Donaldson, Leo Zeef and Peter Briggs, Manchester University Bioinformatics Core
	Facility</p>
    
      <h2>Contact</h2>
      <p>Peter Briggs (peter.briggs@manchester.ac.uk)</p>
    
      <div class="box">
      <h2>For developers</h2>
      <p>See the source code on GitHub: <a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator">https://github.com/fls-bioinformatics-core/RnaChipIntegrator</a></p>

      <p>You can clone the project with <a href="http://git-scm.com">Git</a> by running:
      <pre>$ git clone git://github.com/fls-bioinformatics-core/RnaChipIntegrator</pre>
      </p>
      </div>

  </div>

</body>
</html>
