<!DOCTYPE html>
<html>
<head>
  <meta charset='utf-8'>

  <title>fls-bioinformatics-core/RnaChipIntegrator @ GitHub</title>

  <style type="text/css">
    body {
      margin-top: 1.0em;
      background-color: #ffffff;
      font-family: Helvetica, Arial, FreeSans, san-serif;
      color: #000000;
    }
    #container {
      margin: 0 auto;
      padding: 20px;
      width: 800px;
      background: #ffffff;
    }
    h1 { font-size: 3.0em; color: #000000; margin-bottom: 3px; }
    h1 .small { font-size: 0.4em; }
    h1 a { text-decoration: none }
    h2 { font-size: 1.2em; color: #000000; }
    h3 { color: #000000; font-size: 1.0em; font-style: italic;}
    a { color: #000000; }
    .description { font-size: 1.1em; margin-bottom: 20px; margin-top: 10px; font-style: italic;}
    .download { float: right; }
    .important { background: #fcc; font-style: italic; }
    pre { background: #000; color: #fff; padding: 15px;}
    hr { border: 0; width: 80%; border-bottom: 1px solid #aaa}
    .footer { text-align:center; padding-top:30px; font-style: italic; }
    .box { padding:10px; background: #ddd; }
    .code { font-family: Courier, monospace; }
  </style>
</head>

<body>

  <div id="container">

    <h1><a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator">RnaChipIntegrator</a></h1>

    <div class="description">
      Integrative analyses of RNA-Seq with ChIP-Seq data
    </div>

    <div class="box">
    <h2>About</h2>
      <p>RnaChipIntegrator is a utility that performs integrated analyses of RNA-Seq and ChIP-Seq
	data, identifying the nearest ChIP peaks to each transcript, and vice versa. It can also be
	used with any set of genomic features e.g. canonical genes, CpG islands etc, or expression
	data e.g. microarrays.</p>
    </div>
    
    <h2>Download</h2>

      <p>Download the latest version:</p>

      <ul>
	<li><a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator/archive/v0.4.0.tar.gz">RnaChipIntegrator-0.4.0.tar.gz</a></li>
	<li><a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator/archive/v0.4.0.zip">RnaChipIntegrator-0.4.0.zip</a></li>
      </ul>

       <p>See the <a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator/blob/master/ChangeLog">change log</a> for the most recent changes, and the <a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator/downloads">downloads page</a> for all available versions.</p>

    <div class="important">
      <p>This release addresses a bug has been discovered in the previous version of the
	code (v0.3.3), which meant that peaks within transcripts or promoter regions on
	the negative strand might sometimes be incorrectly flagged as not within those
	regions.</p>
      <p>This bug did <b>not</b> affect the reporting of the peaks and transcripts</p>
    </div>
    
    <h2>Installation</h2>

    <p>A number of installation options are available:</p>

    <ol>
      <li><p>You can use Python's setuptools to install the RnaChipIntegrator code. To do this,
	  download the archive (either .tar.gz or .zip), unpack it and install as follows:</p>

	  <pre>$ cd RnaChipIntegrator-0.4.0
$ sudo python setup.py install</pre></li>

      <li><p>Alternatively if you're using pip (see
	  <a href="http://www.pip-installer.org/en/latest/index.html">www.pip-installer.org/en/latest/index.html</a>) to manage your Python installations then you can install as follows:</p>

      <pre>$ sudo pip install RnaChipIntegrator.tar.gz</pre></li>

      <li><p>If you don't want to do a system install then you can just unpack the archive and run
	  the Python code directly using:</p>
	
	<pre>$ /path/to/RnaChipIntegrator.py [options] ...</pre>

      <p>(Note you will need to add the '.py' extension to the program name when running the
	examples below.)</p></li>
    </ol>

    <p>See the <a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator/blob/master/INSTALL">INSTALL</a> document for more detailed information about the how to install and uninstall the utility.<p>

    <h2>Usage</h2>

      <p>General usage syntax is:</p>

      <pre>$ RnaChipIntegrator [options] &lt;rna-data-file&gt; &lt;chip-data-file&gt;</pre>

      <p>Use <span class="code">-h</span> to produce a list of options and descriptions of their
	functions:</p>

      <pre>$ RnaChipIntegrator -h</pre>

      <p>Specific examples using example data included in the downloads:</p>

      <pre>$ RnaChipIntegrator --window=130000 ExpressionData.txt ChIP_edges.txt
$ RnaChipIntegrator --window=130000 ExpressionData.txt ChIP_peaks.txt</pre>

      <p>See the <a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator/blob/master/doc/MANUAL.markdown">manual</a> for more detailed information about the analyses.<p>

      <h3>Expression Data</h3>

      <p>The expression data (RNA-seq) file must be a tab-delimited file with 5 columns of data
	for each genomic feature (one per line):</p>

      <pre>ID  chr  start  end  strand</pre>

      <p><span class="code">chr</span> is the chromosome name, <span class="code">start</span> and
	<span class="code">end</span> define the limits of the feature, and
	<span class="code">strand</span> must be either + or -. <span class="code">GeneID</span>
	is a name which is used to identify the genomic feature in the output. Genes in between a
	reported gene and the ChIP site are also reported.</p>

      <p>Optionally there can be a 6th column, indicating whether the gene was differentially
	expressed (= 1) or not (= 0).</p>

      <p>(See the example <a href="https://raw.github.com/fls-bioinformatics-core/RnaChipIntegrator/master/examples/ExpressionData.txt">ExpressionData.txt</a> file.)</p>

      <h3>ChIP-Seq Data</h3>

      <p>The ChIP-seq data file must be a tab-delimited file with 3 columns of data for each
	ChIP peak (one per line):</p>

      <pre>chr  start  end</pre>

      <p><span class="code">chr</span> is the chromosome name (must match those in the expression
	data file), and <span class="code">start</span> and <span class="code">end</span>
	define the ChIP peaks - these can either be summits (in which case
	<span class="code">end</span> - <span class="code">start</span> = 1), or regions (with
	 <span class="code">start</span> and <span class="code">end</span> indicating the
	 extent).</p>

      <p>(See the example <a href="https://raw.github.com/fls-bioinformatics-core/RnaChipIntegrator/master/examples/ChIP_edges.txt">ChIP_edges.txt</a> and <a href="https://raw.github.com/fls-bioinformatics-core/RnaChipIntegrator/master/examples/ChIP_peaks.txt">ChIP_peaks.txt</a> files.)</p>

      <p>Note that different analyses will be selected depending on whether the ChIP peaks are
	defined as summits or regions.</p>
    
      <h2>Dependencies</h2>
      <p>RnaChipIntegrator was written for Python 2.7, but has also been used with Python
	2.4. It uses the xlwt, xlrd and xlutils Python libraries to write its XLS output files.
	If these are not available then the program will still run but won't produce an XLS
	file.</p>
    
      <h2>License</h2>
      <p>Artistic Licence 2.0</p>
    
      <h2>Authors</h2>
      <p>Ian Donaldson, Leo Zeef and Peter Briggs, University of Manchester, <a href="http://www.ls.manchester.ac.uk/research/facilities/#bioinf">Faculty of Life Sciences Bioinformatics Core Facility</a>
	Facility</p>
    
      <h2>Contact</h2>
      <p>Peter Briggs (peter.briggs [at] manchester.ac.uk)</p>
    
      <div class="box">
      <h2>For developers</h2>
      <p>See the source code on GitHub: <a href="https://github.com/fls-bioinformatics-core/RnaChipIntegrator">https://github.com/fls-bioinformatics-core/RnaChipIntegrator</a></p>

      <p>You can clone the project with <a href="http://git-scm.com">Git</a> by running:
      <pre>$ git clone git://github.com/fls-bioinformatics-core/RnaChipIntegrator</pre>
      </p>
      </div>

  </div>

</body>
</html>
